criu/apparmor.c

```diff
@@ -0,0 +1,817 @@
+		return NULL;
+	}
+
+	n = write(fd, policy, policy_len);
+	close(fd);
+	if (n < 0 || n != policy_len) {
+		pr_perror("couldn't write policy for %s", file);
+		return NULL;
+	}
+
+	n = run_command(cachedir, sizeof(cachedir), apparmor_cache_exec, NULL);
+	if (n < 0) {
+		pr_err("apparmor parsing failed %d\n", n);
+		return NULL;
+	}
+
+	n = run_command(NULL, 0, apparmor_parser_exec, &args);
+	if (n < 0) {
+		pr_err("apparmor parsing failed %d\n", n);
+		return NULL;
+	}
+
+	n = snprintf(file, sizeof(file), "%s/cache/%s/%s", policydir, cachedir, clean_name);
+	if (n < 0 || n >= sizeof(policy)) {
+		pr_err("policy name %s too long\n", clean_name);
+		return NULL;
+	}
+
+	fd = open(file, O_RDONLY);
+	if (fd < 0) {
+		pr_perror("couldn't open %s", file);
+		return NULL;
+	}
+
+	if (fstat(fd, &sb) < 0) {
+		pr_perror("couldn't stat fd");
+		goto out;
+	}
+
+	ret = mmap(NULL, sb.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
+	if (ret == MAP_FAILED) {
+		pr_perror("mmap of %s failed", file);
+		goto out;
+	}
+
+	*len = sb.st_size;
+out:
+	close(fd);
+	return ret;
+}
+
+#define NEXT_AA_TOKEN(pos)					\
+	while (*pos) {						\
+		if (*pos == '/' &&				\
+				*(pos+1) && *(pos+1) == '/' &&	\
+				*(pos+2) && *(pos+2) == '&') {	\
+			pos += 3;				\
+			break;					\
+		}						\
+		if (*pos == ':' &&				\
+				*(pos+1) && *(pos+1) == '/' &&	\
+				*(pos+2) && *(pos+2) == '/') {	\
+			pos += 3;				\
+			break;					\
+		}						\
+		pos++;						\
+	}
+
+static int write_aa_policy(AaNamespace *ns, char *path, int offset, char *rewrite, bool suspend)
+{
+	int i, my_offset, ret;
+	char *rewrite_pos = rewrite, namespace[PATH_MAX];
+
+	if (rewrite && suspend) {
+		pr_err("requesting aa rewriting and suspension at the same time is not supported\n");
+		return -1;
+	}
+
+	if (!rewrite) {
+		strncpy(namespace, ns->name, sizeof(namespace) - 1);
+	} else {
+		NEXT_AA_TOKEN(rewrite_pos);
+
+		switch(*rewrite_pos) {
+		case ':': {
+			char tmp, *end;
+
+			end = strchr(rewrite_pos+1, ':');
+			if (!end) {
+				pr_err("invalid namespace %s\n", rewrite_pos);
+				return -1;
+			}
+
+			tmp = *end;
+			*end = 0;
+			strlcpy(namespace, rewrite_pos+1, sizeof(namespace));
+			*end = tmp;
+
+			break;
+		} default:
+			strlcpy(namespace, ns->name, sizeof(namespace));
+			for (i = 0; i < ns->n_policies; i++) {
+				if (strcmp(ns->policies[i]->name, rewrite_pos))
+					pr_warn("binary rewriting of apparmor policies not supported right now, not renaming %s to %s\n", ns->policies[i]->name, rewrite_pos);
+			}
+		}
+	}
+
+	my_offset = snprintf(path+offset, PATH_MAX-offset, "/namespaces/%s", ns->name);
+	if (my_offset < 0 || my_offset >= PATH_MAX-offset) {
+		pr_err("snprintf'd too many characters\n");
+		return -1;
+	}
+
+	if (!suspend && mkdir(path, 0755) < 0 && errno != EEXIST) {
+		pr_perror("failed to create namespace %s", path);
+		goto fail;
+	}
+
+	for (i = 0; i < ns->n_namespaces; i++) {
+		if (write_aa_policy(ns, path, offset + my_offset, rewrite_pos, suspend) < 0)
```

Comment:
Recursion again... Can we use DFS over tree like we normally do. (see mnt_subtree_next() for an example)